### PR TITLE
Add check for non-empty data vector in new function

### PR DESCRIPTION
**Pull request by joda01**
Add a check to ensure the data vector is not empty before proceeding.

## 📝 Description

Fixes: #694

**Important**:
- Ensure you are assigned to the linked issue before submitting this PR
- This PR should strictly implement what the linked issue describes
- Do not include changes beyond the scope of the linked issue

---

## 🛠️ Changes Made
- [x] Check for empty data vector when creating a new image

---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** Added `test_image_from_empty_vec()` 
- [x] **Manual Verification:** Just execute the peace of unit test
- [x] **Performance/Edge Cases:** Not applicable

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [x] 🟢 **No AI used.**
- [ ] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [ ] The linked issue has been approved by a maintainer
- [x] This PR strictly implements what the linked issue describes (no scope creep)
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

## 💭 Additional Context
Add any other context or screenshots about the pull request here.

### CODE REVIEW

**Review by sidd-27 — APPROVED** [review]

File: crates/kornia-image/src/image.rs [hunk 1]
```diff
@@ -698,6 +703,21 @@ mod tests {
...
+    fn test_image_from_empty_vec() -> Result<(), ImageError> {
+        let image: Image<f32, 3, CpuAllocator> = Image::new(
+            ImageSize {
+                height: 0,
+                width: 0,
+            },
+            vec![0.0; 0],
+            CpuAllocator,
+        );
+        assert!(image.is_err(), "Image::new should have failed with empty vec");
```
[the remark this answers is not in the document]
Comment (joda01) [reply crates/kornia-image/src/image.rs]: Corrected assignment data type to `Result<Image<f32, 1, CpuAllocator>, ImageError> `

**Comment by sidd-27** [comment]
fix lint tests by running `cargo fmt`

**Comment by joda01** [comment]
Fixes #694

**Comment by edgarriba** [comment]
hey, this has merge conflicts — can you rebase on main?

**Comment by joda01** [comment]
@edgarriba  I rebased on main but now have lots of file changes. Did I made a mistake?

**Comment by joda01** [comment]
@edgarriba I rebased and fixed the merge conflicts.

**Comment by edgarriba** [comment]
Hey @joda01 — thanks for rebasing! 

Yes, a rebase on an active repo can sometimes show extra file changes if the branch was behind by many commits — that's normal and expected. As long as your actual changes are only in `crates/kornia-image/src/image.rs`, you're good.

However, looking at the automated review feedback, there are a couple of compilation issues that need to be fixed before this can be merged:

1. **Error type issue**: The empty-vector guard uses `Err(ImageDataNotInitialized())`, but `ImageDataNotInitialized` is an `ImageError` unit enum variant (not callable with `()`). It should be `Err(ImageError::ImageDataNotInitialized)` or a similar correct variant.

2. **Test type mismatch**: The test declares `image: Image<...>` but assigns a `Result<...>` — this won't compile. Change it to `let result = Image::new(...)` and then `assert!(result.is_err(),...)`.

3. **Zero-sized image behavior**: The unconditional `data.len() == 0` check also rejects valid 0×0 images. Consider checking `data.len() == 0 && size.width * size.height * C != 0` instead, so a truly empty image (0×0) is still allowed.

Could you fix those up and push an update? @sidd-27 can do the final review and merge once the compile errors are sorted. Thanks for contributing! 🙏

**Comment by joda01** [comment]
Hello @edgarriba ,

thank you for the hint and sorry, the errors came from the rebase I had still fixed them.

To your point 3 which behaviour would you expect? This is exact the actual problem. If a truly empty image is created a seg fault is thrown because a data vector of size 0 is, why ever, not allowed.

The problem occurs here if data has len 0 or is empty.

```rust
        // allocate the image data
        Ok(Self(Tensor3::from_shape_vec(
            [size.height, size.width, C],
            data,
            alloc,
        )?))
```

How should we handle true empty images?
My approach was to just don't allow empty images. But I agree with you probably there is a better solution.
Just let me know what you would expect, I can come up with a solution suggestion afterwars.

Best greetings,
Joachim

**Comment by joda01** [comment]
I dived into the problem and compared a vector of size > 0 and a vector of size = 0.

In storage.rs in function `from_vec`  there is the line

```rust
   let ptr = unsafe { NonNull::new_unchecked(value.as_ptr() as _) };
```
If called with a vector size > 0 a valid pointer address is returned.

vec_not_0

If called with a vector size = 0 an invalid pointer is returned.

vec_0


I think this could be the reason for the segfault.

**Comment by joda01** [comment]
@edgarriba I got it, the problem happens in the `Drop` of the storage

```rust
impl<T, A: TensorAllocator> Drop for TensorStorage<T, A> {
    /// Automatically deallocates the storage's memory when dropped.
    ///
    /// This uses the storage's allocator to properly free the memory.
    fn drop(&mut self) {
        self.alloc
            .dealloc(self.ptr.as_ptr() as *mut u8, self.layout);
    }
}
```

If the data vector is zero size, everything so far is good. But the pointer is just a dangling pointer, since storage has never been allocated. The drop function now tries to dealloc, which fails since no memory was allocated.

So sorry for the confusion. I reverted my change in the image.rs and provided a fundamental fix in the storage.rs, which solves the problem at the root.

**Comment by edgarriba** [comment]
@joda01 excellent root cause analysis! 🎯 You are absolutely right — the `Drop` trying to deallocate a dangling pointer from a zero-length vector is the fundamental issue, not the `Image::new` layer.

Fixing it in `storage.rs` is the correct approach since that is where the allocation (or lack thereof) actually happens. The fix should:
- In `TensorStorage::Drop`, guard against deallocating when the length/capacity is zero (i.e., skip `dealloc` if no memory was actually allocated)

A common pattern is to check `self.layout.size() == 0` or track whether the storage was truly allocated before calling `dealloc`.

Could you push the `storage.rs` fix so we can review it? Once CI is green on that approach, @sidd-27 / @cjpurackal can give the final merge.

**Comment by joda01** [comment]
@sidd-27 I pushed the suggested patch (```crates/kornia-tensor/src/storage.rs```) and executed cargo fmt. However it formatted also the file ```crates/kornia-3d/src/io/pcd/parser.rs```  which I have not touched. 
Please let me know if I made something wrong ...

**Comment by sidd-27** [comment]
> @sidd-27 I pushed the suggested patch (`crates/kornia-tensor/src/storage.rs`) and executed cargo fmt. However it formatted also the file `crates/kornia-3d/src/io/pcd/parser.rs` which I have not touched. Please let me know if I made something wrong ...

i think you need to pull the main branch perhaps? try rebasing once

**Comment by sidd-27** [comment]
fix the end of line checks too

**Comment by joda01** [comment]
@sidd-27 I rebased once. Formating should be clean now.
Relevant change for the bug fix was this line: https://github.com/kornia/kornia-rs/pull/695/changes#diff-9bceae2c8e26efdd86f447b2f7916b10181a76194b8d743d968b0c2797bb6de2R258

I added two new unit tests one in the crates/kornia-tensor/src/storage.rs to check if an empty data vector can be created and dropped without crash and one in the crates/kornia-image/src/image.rs adding the formal missing edge case adding an empty image (data vector size 0, length and width 0).

Please just let me know If there is still something missing.

**Comment by joda01** [comment]
@sidd-27 can you retrigger https://github.com/kornia/kornia-rs/actions/runs/22280720433/job/64451178230?pr=695

I fixed the Fixes in the description

**Comment by edgarriba** [comment]
Hey @joda01 👋 — _(posting on behalf of @edgarriba)_

Almost there! All the Rust/C++/Python tests are passing ✅ — the only failing check is `Validate PR Requirements`. Could you check what that requires? (Might be a label, PR description section, or changelog entry that's missing.)

@sidd-27 has already approved twice, so once that check is green this should be good to merge. Thanks for the contribution!)

**Comment by joda01** [comment]
Hello @edgarriba :) its because of "No linked issue found" I corrected the description yesterday. The pipeline should pass now after a rerun.

Best grettings!
